### ISSUE TITLE
Validation on update and delete

### ISSUE THREAD

**Issue by enj** [kubernetes/kubernetes#59984]
As it stands today, changes to validation logic can break the API if they tighten the requirements for an object.  The general case for this is:

1. Create object `X`
2. Upgrade to new server version
3. Try to update object `X` (add a label for example)
4. The update fails because of validation errors

Specifically, an object already stored within the API is considered invalid.

Another similar issue can occur on a delete which requires an update (i.e. a finalizer needs to be removed).  Since the object cannot be updated in-place, it cannot be deleted without first "fixing" it to match the current validation requirements.

This has the potential to break controllers or any autonomous process.  It makes changes to validation hard to review because you have to understand how preexisting objects will behave under the new validation.  In openshift, this constantly breaks the storage migration process which does no-op updates to all objects in the API.

A possible fix for the no-op update case would be to _not_ run validation when the input and output objects are exactly the same.

For the more complex update cases and the delete case, we could do something similar to `IsOnlyMutatingGCFields` to determine if validation can be safely skipped.

@kubernetes/sig-api-machinery-misc

**Comment by smarterclayton** [comment]
I think that makes sense.  However, is the no-op case really special for update?  Any other update (including controllers) would be broken. 

`IsOnlyMutatingGCFields` is a bit of a hack - is it the exception rather than the rule?

**Comment by sttts** [comment]
> Specifically, an object already stored within the API is considered invalid.

Tightened validation and existing invalid objects is a serious problem in every cluster. The only question is whether we keep going during migration and graceful/with-initializer deletion, or we fail early and hard.

> is the no-op case really special for update? Any other update (including controllers) would be broken.

I tend to prefer "fail early" and answer that it is not special. Instead, we should delete the object from the system ASAP, possibly with a good migration UI that warns or asks for confirmation. But as controllers are broken otherwise, allowing mirgration just postpones the error to later and it will be much harder to debug and understand why objects stops behaving as expected.

**Comment by enj** [comment]
> we fail early and hard

Considering this stuff fails upgrades for our clusters regularly as it is, failing hard is unacceptable IMO.

> we should delete the object from the system ASAP, possibly with a good migration UI that warns or asks for confirmation

That is not how I see the problem at all - deleting data is not a solution.

I consider an object stored in the API to be, by definition, valid.  Thus validation does not check "is the object valid?", it checks if "the change to the object is valid."

`any object already stored in the API` + `any valid change to that object` == `no validation error` (you only validate the entire object on create)

**Comment by sttts** [comment]
> That is not how I see the problem at all - deleting data is not a solution.

Thinking more about a big, bold warning about invalid objects at the very least, so the cluster admin can act accordingly, potentially deleting the objects or fixing them. If those warnings are ignored, later object updates will fail, without a clear connection to the recent cluster upgrade. Not sure that's a situation we want our users in.

> I consider an object stored in the API to be, by definition, valid. Thus validation does not check "is the object valid?", it checks if "the change to the object is valid."

Differential checks are much harder to do and prone to (programming) errors. In theory I agree: perfect differential validation would be good and solve the problem. In practice, we have a mixture of differential and absolute validations, by far more of the later. So we don't have much choice if we want a solution now.

**Comment by deads2k** [comment]
Finalizer and controller (status) updates should be fixed using subresources that only validate the part they are mutating.  Many status validations already do this properly

**Comment by lavalamp** [comment]
cc @jpbetz @lavalamp

**Comment by gnufied** [comment]
I do not think this is a 0 or 1 situation. Some objects may have live resources(such as pods that are running) associated with it and some objects will be in pending or other such state. 

I believe that - validations that tighten API spec to prevent pending or other zombie state should be allowed. Is it possible that - this is more of a tooling problem for such cases? For example - if a validation prevents creation of objects that will not run or can't be used, then the upgrade tool should check for such API objects and allow admin to delete them.  For example - a `pending` pod or a `pending` PVC can be deleted without much consequence. 

Also discussed offline a bit with @deads2k - and it appears that there is no consistent support for finalizers as a subresource yet, so any tightening of validation will cause such objects to be undeletable.
